### ISSUE TITLE
Cannot use JS strings without copying them into wasm memory

### ISSUE THREAD

**Issue by fitzgen** [rustwasm/wasm-bindgen#287]
This doesn't work, but it should:

```rust
#[wasm_bindgen]
extern {
    #[wasm_bindgen(js_name = String)]
    pub extern type JsString;

    pub fn slice(this: &JsString, start: u32, end: u32) -> JsString;
}
```

This is blocking all the `String` related bindings in #275.

I haven't dug in very far, but I think the issue is that strings are special cased by wasm-bindgen to always copy in/out of wasm memory, and so we can't reference them as an opaque extern type.

**Comment by kzvi** [comment]
Aside from `js_name` not working, this works for me. Here is an example test case of it working (see `build.sh`): https://github.com/kzvi/wbg_test

**Comment by kzvi** [comment]
[This line](https://github.com/rustwasm/wasm-bindgen/blob/9e01e67aa3cb94493372f88acb4c8e5b2b0a684e/crates/backend/src/ast.rs#L387) in crates/backend/src/ast.rs makes the assumption that the literal type name of a method in Rust is going to be the same as the type name of a method in Javascript. This is why this is breaking.

In order to fix this, either:

- Every method declaration needs to be annotated with the JS type name, e.g. `#[wasm_bindgen(method, js_class = String)]`
- The code in the "backend" crate would need to have some mutable state to keep track of which class methods are being defined for.

**Comment by ohanar** [comment]
I believe this was resolved in #295.
